casper/src/main/scala/coop/rchain/casper/Validate.scala

```diff
@@ -87,6 +89,40 @@ object Validate {
     }
   }
 
+  // This is not a slashable offence
+  def timestamp[F[_]: Monad: Log: Time](
+      b: BlockMessage,
+      dag: BlockDag): F[Either[RejectableBlock, IncludeableBlock]] =
+    for {
+      currentTime  <- Time[F].currentMillis
+      timestamp    = b.header.get.timestamp
+      beforeFuture = currentTime >= timestamp
```

Comment:
I think this should be `currentTime + drift >= timestamp`, where `drift` is a constant which represents the amount of time another clock could be ahead of yours. We should probably make this configurable, but for now can probably just be a constant in the class. 

We should also probably set a minimum time that we accept; something like `recent = timestamp >= currentTime - delay`, where `delay` is another constant which represents how long we expect a message takes to reach us, i.e. an explicit synchrony constraint.

Reply:
I'm worried that this allows attacks to intentionally set their block timestamps to be in the future and then that prevents the next validator in the round robin from creating a valid timestamp that is after that future timestamp before their allotted time is up.

Reply:
Did you mean

`recent = timestamp <= currentTime - delay` ?

Reply:
Oh wow geth does this

```
allowedFutureBlockTime = 15 * time.Second
...
if header.Time.Cmp(big.NewInt(time.Now().Add(allowedFutureBlockTime).Unix())) > 0 {
			return consensus.ErrFutureBlock
		}
	}
	if header.Time.Cmp(parent.Time) <= 0 {
		return errZeroBlockTime
}
```
from https://github.com/ethereum/go-ethereum/blob/d429a92f09e476c431830cef48690dc3784153c7/consensus/ethash/consensus.go#L234

Essentially strictly checks that block timestamp is before parent timestamp and for checking for future blocks it adds a `drift`.

I'm still a bit worried though; ethereum PoW doesn't suffer the same round robin attack I've outlined above.

Reply:
I guess we can allow a drift that is half the allotted time for round robin?